src/PartitionGraph.f90

```diff
@@ -1703,28 +1720,42 @@ SUBROUTINE SpatialKernighanLin_PartitionGraph(thisGraph,L1,L2)
 !-------------------------------------------------------------------------------
 !> @brief Calculate metrics relevant to the quality of partition
 !> @param thisGraph the partitioned graph to check metrics of
+!> @param maxnsr the maximum number of source regions
 !> @param mmr the ratio of maximum sized(weighted) group to minimum
 !> @param srms the rms of the group size(weighted) difference from optimal
 !> @param ecut the total weight of edges cut
 !> @param comm the total weight of communication between groups
 !>
-    SUBROUTINE calcDecompMetrics_PartitionGraph(thisGraph,mmr,srms,ecut,comm)
+    SUBROUTINE calcDecompMetrics_PartitionGraph(thisGraph,maxnsr,mmr,srms,ecut,comm)
       CHARACTER(LEN=*),PARAMETER :: myName='calcDecompMetrics'
       CLASS(PartitionGraphType),INTENT(IN) :: thisGraph
-      REAL(SRK),INTENT(OUT) :: mmr,srms,ecut,comm
+      REAL(SRK),INTENT(OUT) :: maxnsr,mmr,srms,ecut,comm
       INTEGER(SIK) :: ig,igstt,igstp,in,ineigh,iv,ivert,neighGrp
-      REAL(SRK) :: wtSum,wtGrp,lgroup,sgroup,optSize,wtDif
+      REAL(SRK) :: wtSum,wtGrp,lgroup,sgroup,optSize,wtDif,gwt
       INTEGER(SIK),ALLOCATABLE :: grpMap(:),uniqueGrps(:)
 
+      maxnsr=0.0_SRK
       mmr=0.0_SRK
       srms=0.0_SRK
       ecut=0.0_SRK
       comm=0.0_SRK
       IF(ALLOCATED(thisGraph%groupIdx)) THEN
+        !Compute the max number of source regions
+        DO ig=1,thisGraph%nGroups
+          igstt=thisGraph%groupIdx(ig)
+          igstp=thisGraph%groupIdx(ig+1)-1
+          gwt=0.0_SRK
+          DO iv=igstt,igstp
+            gwt=gwt+thisGraph%wts(thisGraph%groupList(iv))
+          ENDDO
+          maxnsr=MAX(maxnsr,REAL(gwt,SRK))
+        ENDDO
+        maxnsr=maxnsr/SUM(thisGraph%wts)
+
```

Comment:
Might be reasonable to put a DBC statement at the end to verify that 
`(maxnsr > 0_SRK) .AND. (maxnsr < 1.0_SRK)`

Reply:
Good call.

Reply:
While I'm doing the one, are these reasonable for the others?
```
      ENSURE((maxnsr > 0.0_SRK) .AND. (maxnsr <= 1.0_SRK))                                  
      ENSURE((mmr > 0.0_SRK)                                                                                                                                                             
      ENSURE((srms > 0.0_SRK) .AND. (srms <= 100.0_SRK))                                    
      ENSURE(ecut >= 0.0_SRK)                                                               
      ENSURE(comm >= 0.0_SRK)            
```